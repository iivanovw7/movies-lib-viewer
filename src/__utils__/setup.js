/**
 * Module contains tests setup
 * @module __utils__/setup
 * @author Igor Ivanov
 */
import 'core-js/stable';
import 'regenerator-runtime/runtime';
import Enzyme, { configure } from 'enzyme';
import jsdom from 'jsdom';

// eslint-disable-next-line import/no-extraneous-dependencies
const Adapter = require('enzyme-adapter-react-16');

Enzyme.configure({ adapter: new Adapter() });

const dom = new jsdom.JSDOM('<!DOCTYPE html><html><body></body></html>', {
  url: 'http://localhost:4428',
  pretendToBeVisual: true,
});
const win = dom.window;

global.document = win.document;
global.window = win;
global.localStorage = win.localStorage;
global.requestAnimationFrame = win.requestAnimationFrame;
global.cancelAnimationFrame = win.cancelAnimationFrame;

// eslint-disable-next-line no-shadow
global.window.IntersectionObserver = class IntersectionObserver {
  // eslint-disable-next-line no-useless-constructor,require-jsdoc,no-empty-function
  constructor() {}

  // eslint-disable-next-line class-methods-use-this,require-jsdoc
  observe(cb) {
    return cb;
  }

  // eslint-disable-next-line class-methods-use-this,require-jsdoc
  unobserve() {
    return null;
  }
};

global.window.IntersectionObserverEntry = {};
global.window.IntersectionObserverEntry.prototype = {
  intersectionRatio: {},
};

configure({ adapter: new Adapter() });
