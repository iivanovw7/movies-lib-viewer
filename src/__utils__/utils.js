/**
 * Module contains test utils
 * @module __utils__/utils
 * @author Igor Ivanov
 */
import chalk from 'chalk';
import { mount, shallow } from 'enzyme';
import { createMemoryHistory } from 'history';
import React from 'react';
import { createStore } from 'redux';
import configureMockStore from 'redux-mock-store';
import createSagaMiddleware from 'redux-saga';
import { ThemeProvider } from 'styled-components';

import settings from '../config/data';
import rootReducer from '../store/configureStore';

/**
 * Theme mock value.
 * @type {{mode: string}}
 */
const theme = {
  mode: settings.theme,
};

/**
 * Configure saga middleware.
 */
const sagaMiddleware = createSagaMiddleware({});

/* eslint-disable */

/**
 * @param {object} initialState - initial state to store
 * @function storeFactory
 * @returns {Store} Redux store
 */
export const storeFactory = initialState => {
  return createStore(rootReducer, initialState);
};

//Browser history mock
export const history = createMemoryHistory('/');

//Store mock
export const mockStore = configureMockStore([sagaMiddleware]);

export const testName = (title, description) => `[${chalk.yellow(title)}]: ${description}`;

/**
 * Styled theme mount HOC
 * @param child
 * @return {*}
 */
export function mountWithTheme(child) {
  return mount(child, {
    wrappingComponent: ({ children }) => <ThemeProvider theme={theme}>{children}</ThemeProvider>,
  });
}

/**
 * Styled theme shallow HOC
 * @param child
 * @return {*}
 */
export function shallowWithTheme(child) {
  return shallow(child, {
    wrappingComponent: ({ children }) => <ThemeProvider theme={theme}>{children}</ThemeProvider>,
  });
}

/* eslint-enable */
