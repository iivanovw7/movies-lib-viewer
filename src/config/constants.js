/**
 * Module contains main application constants.
 * @module config/constants
 */

// Type keys for Logger messages
export const LOGGER_WARNING = Symbol('Warning');
export const LOGGER_STORE = Symbol('Store');
export const LOGGER_ERROR = Symbol('Error');
export const LOGGER_INFO = Symbol('Info');
export const LOGGER_SUCCESS = Symbol('Success');

// http method types map
export const HTTP_METHOD_GET = 'GET';
export const HTTP_METHOD_POST = 'POST';
export const HTTP_METHOD_HEAD = 'HEAD';
export const HTTP_METHOD_DELETE = 'DELETE';
export const HTTP_METHOD_PUT = 'PUT';
export const HTTP_METHOD_PATCH = 'PATCH';

// Themes
export const LIGHT_THEME = 'light';
export const DARK_THEME = 'dark';
