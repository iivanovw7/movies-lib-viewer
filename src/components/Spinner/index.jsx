/**
 * Module contains Spinner component
 * @module components/Spinner
 * @author Igor Ivanov
 */
import * as PropTypes from 'prop-types';
import React from 'react';
import { SelfBuildingSquareSpinner } from 'react-epic-spinners';
import { connect } from 'react-redux';
import { createSelector } from 'reselect';

import { colorSet } from '../../config/styles';
import { makeSelectTheme } from '../../containers/ThemeSwitch/model/selectors';

/**
 * Default spinner size in `px`.
 * @type {number}
 */
export const defaultSpinnerSize = 50;

/**
 * Creates image component.
 * @method
 * @param {Object} props
 *  contains component props
 *  @see {@link module:components/Spinner~Spinner.propTypes}
 * @return {Node} React component with children.
 * @constructor
 */
const Spinner = ({ color, size, style, mode }) => (
  <SelfBuildingSquareSpinner color={color || colorSet[mode].textColorPrimary} size={size} style={style} />
);

/**
 * @name Spinner.propTypes
 * @type {Object}
 * @param {Object} props - React PropTypes
 * @property {number} props.size - specifies how large the spinner should be rendered.
 * @property {string} [props.color = transparent] - Specifies the color of the spinner.
 * @return {Array} React propTypes
 */
Spinner.propTypes = {
  mode: PropTypes.string.isRequired,
  size: PropTypes.number,
  color: PropTypes.string,
  style: PropTypes.object,
};

Spinner.defaultProps = {
  size: defaultSpinnerSize,
  style: {
    position: 'absolute',
    top: '50%',
    left: '50%',
    transform: `translate(-${defaultSpinnerSize * 0.5}px, -${defaultSpinnerSize * 0.5}px)`,
  },
};

/**
 * Function selects parts of the state required in component.
 * @method
 * @see {@link module:containers/App/model/selectors}
 * @return {Function} selector
 */
const mapStateToProps = createSelector(makeSelectTheme(), (mode) => ({
  mode,
}));

export default connect(mapStateToProps, null)(Spinner);

export { Spinner as OriginalSpinner };
