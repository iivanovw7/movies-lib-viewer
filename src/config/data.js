/**
 * Module contains main application configuration.
 * @module config/styles
 */

import { DARK_THEME } from './constants';

const settings = {
  /**
   * Log level, can be set to below options:
   *  - error [default, only errors]
   *   - debug [all levels]
   *   - off   [no logging]
   * @type {string}
   */
  logLevel: 'error',
  /**
   * Theme config.
   * @type {string}
   */
  theme: DARK_THEME,
  /**
   * Locale config.
   * @type {string}
   */
  locale: 'en',
  /**
   * Network config.
   * @type {Object}
   */
  net: {
    /**
     * Base the movie db API url.
     * @type {string}
     */
    tmdbUrl: 'https://api.themoviedb.org/3',
    /**
     * Base the movie db API image url.
     * @type {string}
     */
    tmdbImageUrl: 'https://image.tmdb.org',
    /**
     * Default request timeout.
     * @type {number}
     */
    requestTimeout: 10000,
  },
};

/**
 *  Sets application debugger in 'errors only' mode if started in production (development) mode.
 *
 *  @param {object} object
 *      initial settings.
 *  @param {string} mode
 *      application mode, defined during build by webpack.
 *  @returns {void} - adds modifications before exporting variable.
 *
 */
(function merge(object, mode) {
  const config = {};

  if (mode === 'test') {
    config.logLevel = 'off';
  }
  if (mode === 'development') {
    config.logLevel = 'debug';
  }

  Object.assign(object, config);
  // eslint-disable-next-line no-undef
})(settings, CONFIG);

export default settings;
