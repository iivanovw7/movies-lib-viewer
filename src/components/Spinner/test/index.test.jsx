/**
 * Module contains Spinner component test suite
 */
import { mount, shallow } from 'enzyme';
import React from 'react';
import { SelfBuildingSquareSpinner } from 'react-epic-spinners';
import { connect, Provider } from 'react-redux';

import { storeFactory, testName } from '../../../__utils__/common';
import settings from '../../../config/data';
import { colorSet } from '../../../config/styles';
import Spinner, { OriginalSpinner } from '../index';

let undef;
const baseProps = {
  mode: 'dark',
  size: 55,
  color: 'red',
  style: {},
};

const store = storeFactory({
  theme: {
    mode: settings.theme,
  },
});

/* eslint-disable require-jsdoc, react/jsx-props-no-spreading */
// prettier-ignore
describe(testName('Spinner', 'component test suite'), function SpinnerTestSuite() {
  const Composition = (props, shouldMount = false) => {
    if (shouldMount) {
      const Component = connect(null, null)(Spinner);
      const node = (
        <Provider store={store}>
          <Component {...props} />
        </Provider>
      );

      return mount(node);
    }

    return shallow(<OriginalSpinner {...props} />);
  };

  it('Should match snapshot', () => {
    const component = Composition({...baseProps});

    expect(component)
      .toMatchSnapshot();
  });

  it('Should shallow render <Spinner />', () => {
    const component = Composition({...baseProps});

    expect(component.find(SelfBuildingSquareSpinner))
      .toHaveLength(1);
  });

  it('Should shallow render <Spinner /> and check props', () => {
    const component = Composition({...baseProps});

    expect(component.prop('size'))
      .toEqual(baseProps.size);

    expect(component.prop('color'))
      .toEqual(baseProps.color);

    component.setProps({
      size: 108
    });

    expect(component.prop('size'))
      .toBe(108);

    expect(component.prop('style'))
      .toEqual(baseProps.style);


    component.setProps({
      style: undef
    });

    expect(component.prop('style'))
      .toEqual(OriginalSpinner.defaultProps.style);

  });

  it('Should shallow render <Spinner /> and verify color', () => {
    const propsMock ={
      ...baseProps,
      color: undef,
    };
    const component = Composition(propsMock, true);

    expect(component.find('StyledComponent').prop('color'))
      .toEqual(colorSet.dark.textColorPrimary);
  });

});
/* eslint-enable require-jsdoc, react/jsx-props-no-spreading */
