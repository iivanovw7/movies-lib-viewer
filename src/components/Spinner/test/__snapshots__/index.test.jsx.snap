// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[[33mSpinner[39m]: component test suite Should match snapshot 1`] = `
<SelfBuildingSquareSpinner
  animationDuration={6000}
  className=""
  color="red"
  size={55}
  style={Object {}}
/>
`;
